tpi_rs/src/middleware/serial.rs

```diff
@@ -0,0 +1,253 @@
+//! Handlers for UART connections to/from nodes
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use bytes::BytesMut;
+use futures::stream::{SplitSink, SplitStream};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tokio_serial::{DataBits, Parity, SerialPortBuilderExt, SerialStream, StopBits};
+use tokio_util::codec::{BytesCodec, Decoder, Framed};
+
+use super::NodeId;
+
+const OUTPUT_BUF_SIZE: usize = 16 * 1024;
+
+type Sink = SplitSink<Framed<SerialStream, BytesCodec>, BytesMut>;
+type Stream = SplitStream<Framed<SerialStream, BytesCodec>>;
+
+#[derive(Debug)]
+pub struct SerialConnections {
+    handlers: Vec<Handler>,
+}
+
+#[derive(Debug)]
+struct Handler {
+    node: usize,
+    sink: Sink,
+    stream: Arc<Mutex<Stream>>,
+    buf: Arc<Mutex<RingBuffer<OUTPUT_BUF_SIZE>>>,
+    worker: Option<JoinHandle<()>>,
+}
+
+#[derive(Debug)]
+struct RingBuffer<const C: usize> {
+    buf: Vec<u8>,
+    idx: usize,
+    len: usize,
+}
+
+impl SerialConnections {
+    pub fn new() -> Result<Self> {
+        let paths = ["/dev/ttyS2", "/dev/ttyS1", "/dev/ttyS4", "/dev/ttyS5"];
+
+        let mut handlers = vec![];
+
+        for (i, path) in paths.iter().enumerate() {
+            handlers.push(Handler::new(i + 1, path)?);
+        }
+
+        Ok(SerialConnections { handlers })
+    }
+
+    pub fn run(&mut self) {
+        for h in &mut self.handlers {
+            h.start_reader();
+        }
+    }
+
+    pub async fn read(&self, node: NodeId) -> String {
+        let idx = node as usize;
+        let handler = &self.handlers[idx];
+        let buf = handler.buf.lock().await.read();
+
+        String::from_utf8_lossy(&buf).to_string()
+    }
+
+    pub async fn write(&mut self, node: NodeId, data: &[u8]) -> Result<()> {
+        let idx = node as usize;
+
+        self.handlers[idx].write(data).await
+    }
+}
+
+impl Handler {
+    fn new(node: usize, path: &'static str) -> Result<Self> {
+        let baud_rate = 115200;
+        let mut port = tokio_serial::new(path, baud_rate)
+            .data_bits(DataBits::Eight)
+            .parity(Parity::None)
+            .stop_bits(StopBits::One)
+            .open_native_async()?;
+
+        // Disable exclusivity of the port to allow other applications to open it.
+        // Not a reason to abort if we can't.
+        if let Err(e) = port.set_exclusive(false) {
```

Comment:
Does our operation system allow for this? 
This calls for data interference, e.g. 2 processes writing to the serial port simultaneously.
I rather have an API option so users can manually enabled/disable uart communication

Reply:
Yes, and it's intended to be more like 2 processes reading from the serial port simultaneously. We have only one intended user of ports (bmcd) but if someone wants to tap in, they would be free to do so.

Reply:
Disabling exclusivity is the reason why it's possible to open a concurrent, functioning picocom session while bmcd is running. The downside is that bmcd process doesn't seem to receive data (but can send).